### PR TITLE
[posedetection_msgs] Add feature0d_to_image node

### PR DESCRIPTION
**Pull request by wkentaro**
it publish sensor_msgs/Image on topic feature0d_to_image/output

### CODE REVIEW

**Comment by wkentaro** [comment]
screenshot from 2015-04-14 23 54 47

**Comment by garaemon** [comment]
advertise before subscribing topics

**Comment by wkentaro** [comment]
Fixed, thanks.

**Comment by garaemon** [comment]
And add spaces after ,

2015年4月15日水曜日、Kentaro Wada<notifications@github.com>さんは書きました:

> Fixed, thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsk-ros-pkg/jsk_common/pull/862#issuecomment-92902829>
> .
>


-- 
✉︎ from iPhone

**Comment by wkentaro** [comment]
I modified for it, thanks.

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+int main(int argc, char **argv)
+{
+    ros::init(argc, argv, "feature0d_to_image");
+    if( !ros::master::check() )
+        return 1;
+
+    boost::shared_ptr<Feature0DToImage> node(new Feature0DToImage());
+    ros::spin();
+    node.reset();
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: Not needed

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            ROS_WARN("bad frame");
+            return;
+        }
+    }
+};
+
+int main(int argc, char **argv)
+{
+    ros::init(argc, argv, "feature0d_to_image");
+    if( !ros::master::check() )
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: Not needed

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                if( i < msg_ptr->features.orientations.size() ) {
+                    // draw line indicating orientation
+                    CvPoint end_pt = cvPoint(center.x+std::cos(msg_ptr->features.orientations[i])*scale,
+                                             center.y+std::sin(msg_ptr->features.orientations[i])*scale);
+                    cv::line(cv_ptr->image, center, end_pt, CV_RGB(255,0,0));
+                }
+            }
+            _pub.publish(cv_bridge::CvImage(cv_ptr->header, "bgr8", cv_ptr->image));
+        }
+        catch (cv_bridge::Exception error) {
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: `cv_bridge::Exception&`

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    virtual ~Feature0DToImage() {}
+
+    void image_cb(const posedetection_msgs::ImageFeature0DConstPtr& msg_ptr)
+    {
+        cv_bridge::CvImagePtr cv_ptr;
+        try {
+            cv_ptr = cv_bridge::toCvCopy(msg_ptr->image, "bgr8");
+            for(size_t i = 0; i < msg_ptr->features.positions.size()/2; ++i) {
+                float scale = i < msg_ptr->features.scales.size() ? msg_ptr->features.scales[i] : 10.0;
+                CvPoint center = cvPoint(msg_ptr->features.positions[2*i+0], msg_ptr->features.positions[2*i+1]);
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: `cv::Point` instead of `CvPoint` to use opencv2 API

**Comment by wkentaro** [comment]
I modified them.

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    virtual ~Feature0DToImage() {}
+
+    void image_cb(const posedetection_msgs::ImageFeature0DConstPtr& msg_ptr)
+    {
+        cv_bridge::CvImagePtr cv_ptr;
+        try {
+            cv_ptr = cv_bridge::toCvCopy(msg_ptr->image, "bgr8");
+            for(size_t i = 0; i < msg_ptr->features.positions.size()/2; ++i) {
+                float scale = i < msg_ptr->features.scales.size() ? msg_ptr->features.scales[i] : 10.0;
+                cv::Point center = cvPoint(msg_ptr->features.positions[2*i+0], msg_ptr->features.positions[2*i+1]);
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: `cvPoint` is still used, you can replace it with cv::Point constructor.

File: posedetection_msgs/src/feature0d_to_image.cpp [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                cv::circle(cv_ptr->image, center, scale, CV_RGB(0,255,0));
+                if( i < msg_ptr->features.orientations.size() ) {
+                    // draw line indicating orientation
+                    cv::Point end_pt = cvPoint(center.x+std::cos(msg_ptr->features.orientations[i])*scale,
```
Comment (garaemon) [inline posedetection_msgs/src/feature0d_to_image.cpp]: `cvPoint` => `cv::Point`

**Comment by wkentaro** [comment]
Sorry about that, I changed them.

**Comment by wkentaro** [comment]
The travis test fails because of its too long runtime. It exceeds 50minutes.
How can I fix this?

**Comment by garaemon** [comment]
I keep restarting travis.
There should be different reason why testing fails

**Comment by wkentaro** [comment]
Thank you.

**Pull request merged**